Add a chat API

SilverAndro opened the pull request:
Implements / closes #37

Needs some more docs + cleanup, but I think this should be a good overview of how this'll work. Opening as draft until I get the final edges smoothed out. Not particularly *enthused* about outbound client chat specifically, so I might re-do that notably.

SilverAndro committed:
First bit of work, client side is functional!

All inbound events are still on the networking thread, maybe should inject *after* main thread is forced to prevent bugs?

SilverAndro committed:
Server side + mappings bump

SilverAndro committed:
Cleanup mixin injections, more test mod

SilverAndro committed:
Cleanup and clarity, begin work on last message type

SilverAndro committed:
Client side inbound profile independent messages

SilverAndro committed:
Move some stuff around+renames

SilverAndro committed:
HUGE amount of work rewriting this!

SilverAndro committed:
TIL I learned java doesnt have import alias

SilverAndro committed:
and with the power of flex tape, I ~~sawed~~ sealed this type system!

SilverAndro committed:
Merge branch 'QuiltMC:1.19.3' into 1.19.3

SilverAndro committed:
Cleanup, get game launching, more hooks

SilverAndro committed:
More before/after IO hooks

SilverAndro committed:
More cleanup and work, almost at done!

SilverAndro committed:
Apparently that wasnt needed!

SilverAndro committed:
Add some more test:tm:

SilverAndro committed:
Didnt forget these shush

EnnuiL commented on library/management/chat/src/testmod/resources/quilt.mod.json:
```diff
@@ -0,0 +1,32 @@
+    "intermediate_mappings": "net.fabricmc:intermediary",
+    "load_type": "always",
+    "entrypoints": {
+      "init": [
+        "org.quiltmc.qsl.chat.test.ChatApiTest"
+      ]
```
Nitpick
```suggestion
      "init": "org.quiltmc.qsl.chat.test.ChatApiTest"
```

EnnuiL commented on library/management/chat/src/testmod/resources/quilt.mod.json:
```diff
@@ -0,0 +1,32 @@
+{
```
oh heck this might have been the most thorough testmod QMJ that i've ever seen; unfortunately i don't know if this means whenever we should cut the tall poppy or leave it alone :P

SilverAndro replied:
Im pretty sure this is just copied+modified from the client commands one? 😅

EnnuiL replied:
wait heck, it does that; oh dear, we'll make 'em consistent later :P

EnnuiL reviewed (changes requested):
I unfortunately can't review the entire API today, but do note that I'm personally excited for it! For now, here are some nitpicks to the testmod

SilverAndro committed:
Flatten initializer in testmod qmj

Co-authored-by: Ennui Langeweile <85590273+EnnuiL@users.noreply.github.com>

EnnuiL commented on library/management/chat/src/testmod/java/org/quiltmc/qsl/chat/test/ChatApiTest.java:
```diff
@@ -0,0 +1,47 @@
+public class ChatApiTest implements ModInitializer {
+	@Override
+	public void onInitialize(ModContainer mod) {
+		MixinEnvironment.getCurrentEnvironment().setOption(MixinEnvironment.Option.DEBUG_ALL, true);
+		MixinEnvironment.getCurrentEnvironment().setOption(MixinEnvironment.Option.DEBUG_EXPORT, true);
```
Hm, we really need to integrate this sort of stuff (including a general mixin audit) to QSL itself properly

SilverAndro replied:
....ngl I forgot that was in there, because I couldnt get it to export through a cli flag

If its fine to stay no complaints though

EnnuiL replied:
hm, heck, maybe it'll have to be removed

SilverAndro replied:
Can do in like, 30min 👍

EnnuiL reviewed (changes requested):
oh heck one more note

Leo40Git commented on library/management/chat/src/main/java/org/quiltmc/qsl/chat/api/types/ImmutableAbstractMessage.java:
```diff
@@ -0,0 +1,59 @@
+
+	// These are mostly annotated as @NotNull for dev productivity, I get nice warnings about
+	// What parts of the subclasses I need to implement still
+	abstract public @NotNull EnumSet<QuiltMessageType> getTypes();
+
+	abstract public @NotNull T immutableCopy();
+
+	abstract public @NotNull S asPacket();
```
```suggestion
	public abstract @NotNull EnumSet<QuiltMessageType> getTypes();

	public abstract @NotNull T immutableCopy();

	public abstract @NotNull S asPacket();
```
excuse me ma'am, but it appears your modifiers are on backwards

Leo40Git commented on library/management/chat/src/main/java/org/quiltmc/qsl/chat/api/types/ImmutableS2CProfileIndependentMessage.java:
```diff
@@ -0,0 +1,85 @@
+					return null;
+				})
+		);
+	}
+	public ImmutableS2CProfileIndependentMessage(PlayerEntity player, boolean isOnClientSide, Text message, MessageType.Parameters messageType) {
```
```suggestion
	}

	public ImmutableS2CProfileIndependentMessage(PlayerEntity player, boolean isOnClientSide, Text message, MessageType.Parameters messageType) {
```

Leo40Git commented on library/management/chat/src/main/java/org/quiltmc/qsl/chat/api/types/ImmutableS2CSystemMessage.java:
```diff
@@ -0,0 +1,82 @@
+        this.player = player;
+		this.content = content;
+		this.overlay = overlay;
+	}
+	@Override
```
```suggestion
	}

	@Override
```

Leo40Git commented on library/management/chat/src/main/java/org/quiltmc/qsl/chat/api/types/MutableAbstractMessage.java:
```diff
@@ -0,0 +1,26 @@
+
+import net.minecraft.entity.player.PlayerEntity;
+
+sealed abstract public class MutableAbstractMessage<T extends ImmutableAbstractMessage<T, S>, S> extends ImmutableAbstractMessage<T, S>
```
```suggestion
public abstract sealed class MutableAbstractMessage<T extends ImmutableAbstractMessage<T, S>, S> extends ImmutableAbstractMessage<T, S>
```

Leo40Git commented on library/management/chat/src/main/java/org/quiltmc/qsl/chat/api/types/ImmutableAbstractMessage.java:
```diff
@@ -0,0 +1,59 @@
+ * @param <T> The type of the immutable type
+ * @param <S> The type of the packet form of this class
+ */
+sealed public abstract class ImmutableAbstractMessage<T extends ImmutableAbstractMessage<T, S>, S>
```
```suggestion
public abstract sealed class ImmutableAbstractMessage<T extends ImmutableAbstractMessage<T, S>, S>
```

Leo40Git commented on library/management/chat/src/main/java/org/quiltmc/qsl/chat/api/types/MutableC2SChatMessage.java:
```diff
@@ -0,0 +1,125 @@
+				packet.signature(),
+				packet.messageAcknowledgments()
+		);
+	}
+	public MutableC2SChatMessage(PlayerEntity player, boolean isOnClientSide, String message, Instant timestamp, long salt, @Nullable MessageSignature signature, MessageSignatureList.Acknowledgment messageAcknowledgments) {
```
```suggestion
	}

	public MutableC2SChatMessage(PlayerEntity player, boolean isOnClientSide, String message, Instant timestamp, long salt, @Nullable MessageSignature signature, MessageSignatureList.Acknowledgment messageAcknowledgments) {
```

Leo40Git reviewed (changes requested):
codestyle nitpicks
(1/2 reviews, separating the nitpicks from my one actual concern with the API surface...)

Leo40Git commented on library/management/chat/src/main/java/org/quiltmc/qsl/chat/api/QuiltChatEvents.java:
```diff
@@ -0,0 +1,27 @@
+public final class QuiltChatEvents {
+	public static final ChatApiVoidEvent MODIFY = new ChatApiVoidEvent();
+	public static final ChatApiEvent<Boolean> CANCEL = new ChatApiEvent<>();
+	public static final ChatApiVoidEvent BEFORE_IO = new ChatApiVoidEvent();
+	public static final ChatApiVoidEvent AFTER_IO = new ChatApiVoidEvent();
```
```suggestion
	public static final ChatApiVoidEvent BEFORE_PROCESSING = new ChatApiVoidEvent();
	public static final ChatApiVoidEvent AFTER_PROCESSING = new ChatApiVoidEvent();
```
`IO` means something completely different! "Processing" _works_ for both inbound and outbound messages, but I'd honestly like a second opinion on this.

SilverAndro replied:
Went with "process" instead of processing because the tense felt better /shrug

Leo40Git reviewed (changes requested):
The rest of the API looks good, I'm just concerned about these two event names...

SilverAndro commented:
Yeah I'm not a fan either, if you have suggestions I'm happy to hear em

Was originally just `QuiltChatEvent`, but then the lambdas got all fucky with `Void` types so I had to make a special one

Leo40Git commented:
I was referring to the names of the event _fields_, not the event _class_.
IMO they're fine - though, maybe drop the `Api`?
- `ChatApiEvent<T>` -> `ChatEvent`
- `ChatApiVoidEvent` -> `ChatVoidEvent` (if only we were in C# land and could declare the non-generic `ChatEvent` as a completely different class...)

SilverAndro commented:
Ah sure that makes sense

What dont you like about the field names specifically?

Leo40Git commented:
`IO` has an _extremely_ common meaning that doesn't make sense here (...okay it means input/output which _also_ works but 99% of people will interpret is as _file_ IO).

SilverAndro committed:
Cleanup + resolve

SilverAndro committed:
Un-seal message wrappers, prevent instantiating QuiltChatEvents and InternalQuiltChatApiUtil

SilverAndro commented:
Decided to un-sealed/final the message wrappers to make hacking this API a bit easier

SilverAndro commented:
Just had a thought, would it make sense to change all the wrappers to immutable, and then require `MODIFY` users to return a new wrapper?

Would simplify the code a decent amount compared to the current layout

SilverAndro committed:
Rewrite everything!

Everything is immutable now with `withX` methods for modification. MODIFY now returns a new instance collected from the listeners, or the original if nothing is returned

SilverAndro commented:
> Just had a thought, would it make sense to change all the wrappers to immutable, and then require `MODIFY` users to return a new wrapper?
> 
> Would simplify the code a decent amount compared to the current layout

Yeah so, did that. It is indeed *much* cleaner and easier to manipulate stuff. Pretty happy with this as is now I think other than the lack of docs and naming complaints and pre-signing C2S outbound chat. Overall chopped off about 400 lines of code :shipit:

SilverAndro committed:
Fix events, toString issues, some more docs

SilverAndro committed:
Fix bug with Void chat events add raw chat message handling

SilverAndro committed:
Move events to API, docs, licensing

SilverAndro commented:
Alright think I got this to the point where its good to be reviewed more in depth, so im going to remove the draft designation once the build here succeeds 😄

EnnuiL commented on library/management/chat/src/testmod/java/org/quiltmc/qsl/chat/test/ChatApiTest.java:
```diff
@@ -0,0 +1,79 @@
+
+		final boolean[] didEnableBad = {false};
+		QuiltChatEvents.CANCEL.register(EnumSet.of(QuiltMessageType.CHAT, QuiltMessageType.CLIENT, QuiltMessageType.OUTBOUND), abstractMessage -> {
+			if (abstractMessage instanceof ChatC2SMessage chatC2SMessage) {
+				if (chatC2SMessage.getMessage().equals("!register-bad")) {
+					if (!didEnableBad[0]) {
+						didEnableBad[0] = true;
+						registerBadEvents();
+						return true;
+					} else {
+						return false;
+					}
+				} else {
+					return false;
+				}
+			} else {
+				return false;
+			}
```
Hm, shouldn't this be:
```suggestion
			if (abstractMessage instanceof ChatC2SMessage chatC2SMessage) {
				if (chatC2SMessage.getMessage().equals("!register-bad")) {
					if (!didEnableBad[0]) {
						didEnableBad[0] = true;
						registerBadEvents();
						return true;
					}
				}
			}

			return false
```
...wait, is that what you meant by potential kotlinisms? :P

SilverAndro replied:
...Honestly not sure why I didnt just let that fall through lmao, think I was just tired while writing that

EnnuiL replied:
ah, that's fair enough

EnnuiL commented on library/management/chat/src/testmod/java/org/quiltmc/qsl/chat/test/ChatApiTest.java:
```diff
@@ -0,0 +1,79 @@
+		});
+
+		QuiltChatEvents.MODIFY.register(EnumSet.of(QuiltMessageType.SYSTEM, QuiltMessageType.SERVER, QuiltMessageType.OUTBOUND), abstractMessage -> {
+			if (abstractMessage instanceof SystemS2CMessage systemS2CMessage) {
+				Text content = systemS2CMessage.getContent();
+				if (new Random().nextBoolean()) {
+					return systemS2CMessage.withContent(content.copy().append(Text.literal(", uwu")));
+				} else {
+					return abstractMessage;
+				}
+			} else {
+				return abstractMessage;
+			}
```
Same thing here:
```suggestion
			if (abstractMessage instanceof SystemS2CMessage systemS2CMessage) {
				Text content = systemS2CMessage.getContent();
				if (new Random().nextBoolean()) {
					return systemS2CMessage.withContent(content.copy().append(Text.literal(", uwu")));
				}
			}

			return abstractMessage;
```

EnnuiL commented on library/management/chat/src/testmod/java/org/quiltmc/qsl/chat/test/ChatApiTest.java:
```diff
@@ -0,0 +1,79 @@
+
+		QuiltChatEvents.AFTER_PROCESS.register(EnumSet.allOf(QuiltMessageType.class), System.out::println);
+		QuiltChatEvents.BEFORE_PROCESS.register(EnumSet.allOf(QuiltMessageType.class), message -> {
+			System.out.println(message.getTypes());
```
hm, I'm worried about potential logspam here; I'll have to see once I test this PR in order to give it the `s: tested` seal of approval :P

EnnuiL replied:
Hm, it doesn't seem to be a huge problem

EnnuiL commented on library/management/chat/src/main/java/org/quiltmc/qsl/chat/mixin/client/ClientPlayNetworkHandlerMixin.java:
```diff
@@ -0,0 +1,172 @@
+
+	@Inject(method = "m_fzlgisyq", at = @At(value = "TAIL"))
+	public void quilt$afterOutboundRawChatMessage(String string, CallbackInfo ci) {
+		if (client.player == null) {
+			return;
+		}
```
Hm, maybe?
```suggestion
		if (client.player == null) return;
```
I don't know if others would agree, but I feel like it would be cleaner to inline it here

EnnuiL commented on library/management/chat/src/main/java/org/quiltmc/qsl/chat/mixin/client/ClientPlayNetworkHandlerMixin.java:
```diff
@@ -0,0 +1,172 @@
+
+	@Inject(method = "m_fzlgisyq", at = @At(value = "HEAD"), cancellable = true)
+	public void quilt$cancelAndBeforeOutboundRawChatMessage(String string, CallbackInfo ci) {
+		if (client.player == null) {
+			return;
+		}
```
```suggestion
		if (client.player == null) return;
```

EnnuiL commented on library/management/chat/src/main/java/org/quiltmc/qsl/chat/mixin/client/ClientPlayNetworkHandlerMixin.java:
```diff
@@ -0,0 +1,172 @@
+		}
+
+		RawChatC2SMessage message = new RawChatC2SMessage(client.player, true, string);
+		if (QuiltChatEvents.CANCEL.invoke(message)) {
+			ci.cancel();
+		} else {
+			QuiltChatEvents.BEFORE_PROCESS.invoke(message);
+		}
```
Hm, I feel like after the `ci.cancel()`, a `return;` should be done so you could get rid of the `else` here

EnnuiL commented on library/management/chat/src/main/java/org/quiltmc/qsl/chat/mixin/client/ClientPlayNetworkHandlerMixin.java:
```diff
@@ -0,0 +1,172 @@
+
+	@ModifyVariable(method = "m_fzlgisyq", at = @At("HEAD"), argsOnly = true)
+	public String quilt$modifyOutboundRawChatMessage(String string) {
+		// Not sure *why* this would be null but IDEA is complaining, so, safety first?
```
oh ho ho ho ho ho, i had to deal with so many pains involving null players

EnnuiL commented on library/management/chat/src/main/java/org/quiltmc/qsl/chat/mixin/client/ClientPlayNetworkHandlerMixin.java:
```diff
@@ -0,0 +1,172 @@
+	@ModifyVariable(method = "m_fzlgisyq", at = @At("HEAD"), argsOnly = true)
+	public String quilt$modifyOutboundRawChatMessage(String string) {
+		// Not sure *why* this would be null but IDEA is complaining, so, safety first?
+		if (client.player == null) {
+			return string;
+		}
```
~~Inline it!~~ No don't! This is fine here

EnnuiL commented on library/management/chat/src/main/java/org/quiltmc/qsl/chat/mixin/client/ClientPlayNetworkHandlerMixin.java:
```diff
@@ -0,0 +1,172 @@
+		}
+
+		RawChatC2SMessage message = new RawChatC2SMessage(client.player, true, string);
+		return ((RawChatC2SMessage)QuiltChatEvents.MODIFY.invoke(message, message)).serialized();
```
Hm, a space after a cast looks better
```suggestion
		return ((RawChatC2SMessage) QuiltChatEvents.MODIFY.invoke(message, message)).serialized();
```

EnnuiL commented on library/management/chat/src/main/java/org/quiltmc/qsl/chat/mixin/client/ClientPlayNetworkHandlerMixin.java:
```diff
@@ -0,0 +1,172 @@
+	public void quilt$modifyAndCancelAndBeforeAndAfterOutboundChatMessage(ClientPlayNetworkHandler instance, Packet<?> packet) {
+		if (packet instanceof ChatMessageC2SPacket chatMessageC2SPacket) {
+			ChatC2SMessage message = new ChatC2SMessage(client.player, true, chatMessageC2SPacket);
+			message = (ChatC2SMessage)QuiltChatEvents.MODIFY.invoke(message, message);
```
```suggestion
			message = (ChatC2SMessage) QuiltChatEvents.MODIFY.invoke(message, message);
```

EnnuiL commented on library/management/chat/src/main/java/org/quiltmc/qsl/chat/mixin/client/ClientPlayNetworkHandlerMixin.java:
```diff
@@ -0,0 +1,172 @@
+				instance.sendPacket(message.serialized());
+				QuiltChatEvents.AFTER_PROCESS.invoke(message);
+			}
+		} else {
+			throw new IllegalArgumentException("Received non-ChatMessageC2SPacket for argument to ClientPlayNetworkHandler.sendPacket in ClientPlayNetworkHandler.m_fzlgisyq (sendChatMessage? mapping missing at time of writing)");
+		}
```
Hm, I do wonder if the else should be removed or not

SilverAndro replied:
I left the `else` in for these cases because the control flow felt more english-like to me. "if of type, do x, otherwise throw"

EnnuiL replied:
ah, I see

EnnuiL commented on library/management/chat/src/main/java/org/quiltmc/qsl/chat/mixin/client/ClientPlayNetworkHandlerMixin.java:
```diff
@@ -0,0 +1,172 @@
+
+	@Inject(method = "onProfileIndependentMessage", at = @At(value = "INVOKE", target = "Lnet/minecraft/client/gui/ClientChatListener;m_jytgvbam(Lnet/minecraft/text/Text;Lnet/minecraft/network/message/MessageType$Parameters;)V", shift = At.Shift.AFTER))
+	public void quilt$afterInboundProfileIndependentMessage(ProfileIndependentMessageS2CPacket packet, CallbackInfo ci) {
+		ProfileIndependentS2CMessage message = new ProfileIndependentS2CMessage(client.player, true, packet);
```
Oh, I noticed a lot of these explicit type things; Could you use `var` for those if possible?

EnnuiL commented on library/management/chat/src/main/java/org/quiltmc/qsl/chat/mixin/client/ClientPlayNetworkHandlerMixin.java:
```diff
@@ -0,0 +1,172 @@
+	@ModifyVariable(method = "onProfileIndependentMessage", at = @At(value = "INVOKE", target = "Lnet/minecraft/network/NetworkThreadUtils;forceMainThread(Lnet/minecraft/network/Packet;Lnet/minecraft/network/listener/PacketListener;Lnet/minecraft/util/thread/ThreadExecutor;)V", shift = At.Shift.AFTER), argsOnly = true)
+	public ProfileIndependentMessageS2CPacket quilt$modifyInboundProfileIndependentMessage(ProfileIndependentMessageS2CPacket packet) {
+		ProfileIndependentS2CMessage message = new ProfileIndependentS2CMessage(client.player, true, packet);
+		return (ProfileIndependentMessageS2CPacket)QuiltChatEvents.MODIFY.invoke(message, message).serialized();
```
```suggestion
		return (ProfileIndependentMessageS2CPacket) QuiltChatEvents.MODIFY.invoke(message, message).serialized();
```

EnnuiL commented on library/management/chat/src/main/java/org/quiltmc/qsl/chat/mixin/client/ClientPlayNetworkHandlerMixin.java:
```diff
@@ -0,0 +1,172 @@
+	@Inject(method = "onSystemMessage", at = @At(value = "FIELD", target = "Lnet/minecraft/client/network/ClientPlayNetworkHandler;client:Lnet/minecraft/client/MinecraftClient;", opcode = Opcodes.GETFIELD, ordinal = 1, shift = At.Shift.BEFORE), cancellable = true)
+	public void quilt$cancelAndBeforeInboundSystemMessage(SystemMessageS2CPacket packet, CallbackInfo ci) {
+		SystemS2CMessage message = new SystemS2CMessage(client.player, true, packet);
+		if (QuiltChatEvents.CANCEL.invoke(message) == Boolean.TRUE) {
+			ci.cancel();
+		} else {
+			QuiltChatEvents.BEFORE_PROCESS.invoke(message);
+		}
```
Same situation here; drop the `else` using a `return;` on the cancelling if :P

EnnuiL commented on library/management/chat/src/main/java/org/quiltmc/qsl/chat/mixin/ServerPlayNetworkHandlerMixin.java:
```diff
@@ -0,0 +1,100 @@
+			QuiltChatEvents.BEFORE_PROCESS.invoke(message);
+			instance.sendPacket(message.serialized());
+			QuiltChatEvents.AFTER_PROCESS.invoke(message);
+		} else {
+			throw new IllegalArgumentException("Received non-ChatMessageS2CPacket for argument to ServerPlayNetworkHandler.sendPacket in ServerPlayNetworkHandler.sendChatMessage");
+		}
```
Same "oh no i don't know" situation here :P

EnnuiL commented on library/management/chat/src/main/java/org/quiltmc/qsl/chat/mixin/ServerPlayNetworkHandlerMixin.java:
```diff
@@ -0,0 +1,100 @@
+			ChatS2CMessage message = new ChatS2CMessage(instance.player, false, chatMessageS2CPacket);
+			message = (ChatS2CMessage)QuiltChatEvents.MODIFY.invoke(message, message);
+
+			if (QuiltChatEvents.CANCEL.invoke(message) == Boolean.TRUE) {
+				return;
+			}
+			QuiltChatEvents.BEFORE_PROCESS.invoke(message);
```
Hm, maybe this?
```suggestion
			if (QuiltChatEvents.CANCEL.invoke(message) == Boolean.TRUE) return;

			QuiltChatEvents.BEFORE_PROCESS.invoke(message);
```
If the inlined return ends being bad, then at least the insertion of a new line would be great

EnnuiL commented on library/management/chat/src/main/java/org/quiltmc/qsl/chat/mixin/ServerPlayNetworkHandlerMixin.java:
```diff
@@ -0,0 +1,100 @@
+
+		independentMessage = (ProfileIndependentS2CMessage)QuiltChatEvents.MODIFY.invoke(independentMessage, independentMessage);
+
+		if (QuiltChatEvents.CANCEL.invoke(independentMessage) == Boolean.TRUE) {
+			ci.cancel();
+		} else {
+			QuiltChatEvents.BEFORE_PROCESS.invoke(independentMessage);
+		}
```
Same situation: drop the `else` with the power of `return;` :P

EnnuiL commented on library/management/chat/src/main/java/org/quiltmc/qsl/chat/impl/InternalQuiltChatApiUtil.java:
```diff
@@ -0,0 +1,44 @@
+	private InternalQuiltChatApiUtil() {}
+
+	public static EnumSet<QuiltMessageType> s2cType(QuiltMessageType type, boolean isOnClientSide) {
+		if (isOnClientSide) {
+			return EnumSet.of(type, QuiltMessageType.CLIENT, QuiltMessageType.INBOUND);
+		} else {
+			return EnumSet.of(type, QuiltMessageType.SERVER, QuiltMessageType.OUTBOUND);
+		}
```
oh dear, i don't like this pattern and yet the only real alternative that i have to offer here are `? :`s, which might be worse

SilverAndro replied:
What dont you like about it out of curiosity? It seems like pretty normal code to me

EnnuiL replied:
Excellent question! I don't know! I just don't like it's face! but yeah, compared to other alternatives? yeah, this might be the best one

EnnuiL commented on library/management/chat/src/main/java/org/quiltmc/qsl/chat/api/types/SystemS2CMessage.java:
```diff
@@ -0,0 +1,84 @@
+	}
+
+	@Override
+	public String toString() {
+		final StringBuilder sb = new StringBuilder("SystemS2CMessage{");
+		sb.append("player=").append(player);
+		sb.append(", content=").append(content);
+		sb.append(", overlay=").append(overlay);
+		sb.append(", player=").append(player);
+		sb.append(", isOnClientSide=").append(isOnClientSide);
+		sb.append('}');
+		return sb.toString();
+	}
```
Hm, interesting way of creating the string; I was thinking about `"a" + b + "c"` concatenation, and well? welp, i'm conflicted

SilverAndro replied:
I was getting poked about using concatenation, Ill probably just switch it. One of the few optimizations javac does anyways.

EnnuiL commented on library/management/chat/src/main/java/org/quiltmc/qsl/chat/api/types/ChatC2SMessage.java:
```diff
@@ -0,0 +1,122 @@
+	private final MessageSignatureList.Acknowledgment messageAcknowledgments;
+
+	public ChatC2SMessage(PlayerEntity player, boolean isOnClientSide, ChatMessageC2SPacket packet) {
+		this(
+				player,
+				isOnClientSide,
+				packet.message(),
+				packet.timestamp(),
+				packet.salt(),
+				packet.signature(),
+				packet.messageAcknowledgments()
+		);
```
hm, I don't like the two tabs here, but that's the thing: I don't know if one tab is a standard that is being used here or if it's just Ennui being Ennui here :P
yeah, check other QSL code first, this might just be ennui being ennui here

SilverAndro replied:
`QuiltBuiltinResourcePackProfile` seems to use double tabs at least, so ill leave it

![image](https://user-images.githubusercontent.com/52360088/212959993-2305d02a-8635-440b-8c26-039cef492903.png)

EnnuiL replied:
Ah, I see

EnnuiL commented on library/management/chat/src/main/java/org/quiltmc/qsl/chat/api/types/AbstractChatMessage.java:
```diff
@@ -0,0 +1,70 @@
+	}
+
+	/**
+	 * @return if this message is on the client side or not. This can also be determined through checking for {@link QuiltMessageType#CLIENT}
```
Hm, I think the second sentence would be split off the return? wait oh no this isn't a `{@return}`; Yeah, maybe something better would be:
```suggestion
	 * {@return if this message is on the client side or not}
	 * <p>
	 * This can also be determined through checking for {@link QuiltMessageType#CLIENT}.
```

SilverAndro replied:
![image](https://user-images.githubusercontent.com/52360088/212801645-6a92d026-b9c1-4917-b849-89c4ffa44525.png)
IDEA *realllly* doesnt like that pattern, removing the `{}` seems to work though?
![image](https://user-images.githubusercontent.com/52360088/212801722-39314e2d-25f8-4f0e-8d43-86575a4a6ac5.png)

EnnuiL replied:
wait, it did that all along?
Yeah, the intention is to deduplicate information, although yeah, as an IntelliJ newbie, the raw `@return` does not seem good;
@LambdAurora is this supposed to happen?

EnnuiL commented on library/management/chat/src/main/java/org/quiltmc/qsl/chat/api/types/AbstractChatMessage.java:
```diff
@@ -0,0 +1,70 @@
+	 * For System and Profile Independent messages, the player is the target of the message as players cannot send these to the server.
+	 * <p>
+	 * For Chat messages, on the server inbound, it is the sender of the chat message, while outbound it is the player the message will be sent to.
+	 * Chat messages on the client return the client player.
+	 * @return the {@link PlayerEntity} associated with this packet
```
```suggestion
	 * Chat messages on the client return the client player.
	 *
	 * @return the {@link PlayerEntity} associated with this packet
```

EnnuiL commented on library/management/chat/src/main/java/org/quiltmc/qsl/chat/api/types/AbstractChatMessage.java:
```diff
@@ -0,0 +1,70 @@
+	public abstract @NotNull EnumSet<QuiltMessageType> getTypes();
+
+	/**
+	 * @return this message in its serialized form
```
```suggestion
	 * {@return this message in its serialized form}
```

SilverAndro replied:
![image](https://user-images.githubusercontent.com/52360088/212801804-bca70139-cc72-41f1-80d4-696f9099681d.png)
Same thing as client side detection, reader mode both detects it and makes it raw leading to weird duplication

EnnuiL replied:
Oh no, the duplication is intentional! I just don't know if the `@return` is supposed to be that way or if it was supposed to turn into an english "Return"

SilverAndro replied:
seems like a very odd choice but alright lol

EnnuiL commented on library/management/chat/src/main/java/org/quiltmc/qsl/chat/api/types/AbstractChatMessage.java:
```diff
@@ -0,0 +1,70 @@
+	// What parts of the subclasses I need to implement still
+
+	/**
+	 * @return the set of {@link QuiltMessageType}s associated with this message
```
```suggestion
	 * {@return the set of {@link QuiltMessageType}s associated with this message}
```

EnnuiL commented on library/management/chat/src/main/java/org/quiltmc/qsl/chat/api/types/AbstractChatMessage.java:
```diff
@@ -0,0 +1,70 @@
+import java.util.EnumSet;
+
+/**
+ * An abstract message, extended for the various types of messages.
+ * @param <S> The type of the serialized form of this class, usually a Packet<?> but not required
```
```suggestion
 * An abstract message, extended for the various types of messages.
 *
 * @param <S> The type of the serialized form of this class, usually a Packet<?> but not required
```

EnnuiL commented on library/management/chat/src/main/java/org/quiltmc/qsl/chat/api/ChatEvent.java:
```diff
@@ -0,0 +1,135 @@
+	}
+
+	/**
+	 * @return The result of invoking this event, or ifNull if there are no listeners
```
```suggestion
	 * {@return The result of invoking this event, or ifNull if there are no listeners}
```

EnnuiL commented on library/management/chat/src/main/java/org/quiltmc/qsl/chat/api/ChatEvent.java:
```diff
@@ -0,0 +1,135 @@
+	}
+
+	/**
+	 * @return The result of invoking this event, or null if there are no listeners
```
```suggestion
	 * {@return The result of invoking this event, or null if there are no listeners}
```

EnnuiL reviewed (changes requested):
This PR looks good! I only have feedback about the code style in general; I was going to cover every single instances of things, but yeah, I feel like it would be better if I pointed to the general direction instead;

EnnuiL commented:
wait oh no, I almost completely forgot that the first word is lowercase for `@return`; oh well, if needed, I'll do a second pass of suggestions

SilverAndro committed:
Let testmod conditions fall through

Co-authored-by: Ennui Langeweile <85590273+EnnuiL@users.noreply.github.com>

SilverAndro committed:
More testmod conditional fallthrough

Co-authored-by: Ennui Langeweile <85590273+EnnuiL@users.noreply.github.com>

SilverAndro committed:
Fix testmod from suggestions

SilverAndro committed:
Some doc fixes, `var` and space between cast

SilverAndro committed:
Weird docs but alright

SilverAndro committed:
Return cleanup

SilverAndro committed:
Remove testmod gunk

SilverAndro committed:
More return cleanup

SilverAndro committed:
StringBuilder -> String concat

SilverAndro committed:
minor javadoc fix

EnnuiL commented on library/management/chat/src/main/java/org/quiltmc/qsl/chat/api/types/RawChatC2SMessage.java:
```diff
@@ -0,0 +1,56 @@
+
+	@Override
+	public String toString() {
+		return "RawChatC2SMessage{" + "message='" + message + '\'' +
+				'}';
```
```suggestion
		return "RawChatC2SMessage{" + "message='" + message + '\'' + '}';
```

EnnuiL commented on library/management/chat/src/main/java/org/quiltmc/qsl/chat/api/types/SystemS2CMessage.java:
```diff
@@ -0,0 +1,82 @@
+	}
+
+	public SystemS2CMessage(PlayerEntity player, boolean isOnClientSide, Text content, boolean overlay) {
+        super(player, isOnClientSide);
```
```suggestion
                super(player, isOnClientSide);
```

EnnuiL replied:
wait OH i see the problem: mixed space/tab indentation

SilverAndro replied:
I blame IDEA

EnnuiL commented on library/management/chat/src/main/java/org/quiltmc/qsl/chat/mixin/ServerPlayNetworkHandlerMixin.java:
```diff
@@ -0,0 +1,100 @@
+
+	@Inject(method = "sendProfileIndependentMessage", at = @At("TAIL"))
+	public void quilt$afterProfileIndependentMessage(Text message, MessageType.Parameters parameters, CallbackInfo ci) {
+		ProfileIndependentS2CMessage immutableMessage = new ProfileIndependentS2CMessage(player, false, message, parameters);
```
```suggestion
		var immutableMessage = new ProfileIndependentS2CMessage(player, false, message, parameters);
```

EnnuiL reviewed (approved):
All seems fine now! I only have few more nitpicks, but yeah, that's all, you passed the test (well, at least until I test on prod :P)

SilverAndro committed:
Missed a `var`

Co-authored-by: Ennui Langeweile <85590273+EnnuiL@users.noreply.github.com>

SilverAndro committed:
Hopefully last bit of cleanup

EnnuiL commented on library/management/chat/src/testmod/java/org/quiltmc/qsl/chat/test/ChatApiTest.java:
```diff
@@ -0,0 +1,70 @@
+		final boolean[] didEnableBad = {false};
+		QuiltChatEvents.CANCEL.register(EnumSet.of(QuiltMessageType.CHAT, QuiltMessageType.CLIENT, QuiltMessageType.OUTBOUND), abstractMessage -> {
+			if (abstractMessage instanceof ChatC2SMessage chatC2SMessage) {
+				if (chatC2SMessage.getMessage().equals("!register-bad")) {
```
Heck, one nitpick:
```suggestion
				if (chatC2SMessage.getMessage().equals("!register_bad")) {
```

SilverAndro committed:
_ moment

Co-authored-by: Ennui Langeweile <85590273+EnnuiL@users.noreply.github.com>

OroArmor commented on library/management/chat/src/main/java/org/quiltmc/qsl/chat/api/ChatEvent.java:
```diff
@@ -0,0 +1,135 @@
+	/**
+	 * {@return the result of invoking this event, or ifNull if there are no listeners}
+	 */
+	public R invoke(AbstractChatMessage<?> message, R ifNull) {
```
```suggestion
	public R invokeOrElse(AbstractChatMessage<?> message, R ifNull) {
```

OroArmor commented on library/management/chat/src/main/java/org/quiltmc/qsl/chat/api/ChatEvent.java:
```diff
@@ -0,0 +1,135 @@
+			return ifNull;
+		} else {
+			return result;
+		}
```
```suggestion
		return result != null ? result : ifNull;
```

OroArmor commented on library/management/chat/src/main/java/org/quiltmc/qsl/chat/api/ChatEvent.java:
```diff
@@ -0,0 +1,135 @@
+import java.util.function.BiFunction;
+
+// Cant extend event because the constructor is private, gotta reproduce some of the API surface
+public class ChatEvent<H, R> {
```
Could a common interface for this and `ChatBooleanEvent` be made? if so then that can be api, while these two classes can then move to impl, which i think is better due to how much impl detail there is here

OroArmor commented on library/management/chat/src/main/java/org/quiltmc/qsl/chat/api/QuiltChatEvents.java:
```diff
@@ -0,0 +1,93 @@
+		void handleMessage(AbstractChatMessage<?> abstractMessage);
+	}
+
+	private static ChatEvent.TypedChatApiHook<AbstractChatMessage<?>> modifyToHook(EnumSet<QuiltMessageType> types, Modify modify) {
+		return new ChatEvent.TypedChatApiHook<>() {
+			@Override
+			public EnumSet<QuiltMessageType> getMessageTypes() {
+				return types;
+			}
+
+			@Override
+			public AbstractChatMessage<?> handleMessage(AbstractChatMessage<?> message) {
+				return modify.handleMessage(message);
+			}
+		};
+	}
+
+	private static ChatEvent.TypedChatApiHook<Void> listenToHook(EnumSet<QuiltMessageType> types, Listen modify) {
+		return new ChatEvent.TypedChatApiHook<>() {
+			@Override
+			public EnumSet<QuiltMessageType> getMessageTypes() {
+				return types;
+			}
+
+			@Override
+			public Void handleMessage(AbstractChatMessage<?> message) {
+				modify.handleMessage(message);
+				return null;
+			}
+		};
+	}
```
Big implementation details here, not sure how to move out quite yet.

OroArmor commented on library/management/chat/src/main/java/org/quiltmc/qsl/chat/api/QuiltMessageType.java:
```diff
@@ -0,0 +1,50 @@
+
+	// Directionality
+	INBOUND(QuiltMetaMessageType.DIRECTION),
+	OUTBOUND(QuiltMetaMessageType.DIRECTION);
```
Javadocs for each enum variant would be nice and good clarifaction.

OroArmor commented on library/management/chat/src/main/java/org/quiltmc/qsl/chat/api/types/AbstractChatMessage.java:
```diff
@@ -0,0 +1,74 @@
+ */
+public abstract class AbstractChatMessage<S> {
+	protected final @NotNull PlayerEntity player;
+	protected final boolean isOnClientSide;
```
```suggestion
	protected final boolean isClient;
```
This should go for the rest of the same name.

EnnuiL replied:
Oh, do note that `is` violates QM's conventions :P

OroArmor replied:
https://github.com/QuiltMC/quilt-mappings/blob/d7fa31ee0a3f7620bbaf164e4e0b1058bf77bbec/mappings/net/minecraft/world/World.mapping#L25

EnnuiL replied:
The point still stands:
https://github.com/QuiltMC/quilt-standard-libraries/blob/1.19.3/CONTRIBUTING.md#general

OroArmor replied:
> avoiding the `is` and `has` prefixes when possible

Its not possible to be `cliented`, so its fine in this case.

SilverAndro replied:
I could do `clientSided` though

EnnuiL replied:
oh well, fair enough

OroArmor replied:
im mostly trying to have it match what QM currently uses for clarity

SilverAndro replied:
Makes sense

OroArmor commented on library/management/chat/src/main/java/org/quiltmc/qsl/chat/impl/InternalQuiltChatApiUtil.java:
```diff
@@ -0,0 +1,44 @@
+/**
+ * Provides some shortcuts for common code in the chat api
+ */
+public final class InternalQuiltChatApiUtil {
```
`Util` isnt a great name for a class, maybe `InternalMessageTypeEnumSetBuilder`? (also a bad name :iea:)

OroArmor reviewed (changes requested):
90% of the way there, but quite of bit of impl leaks into the api package

SilverAndro committed:
Attempt to address most concerns

SilverAndro committed:
Ternaries, whoops

SilverAndro committed:
Who needs legal notices anyways smh smh

SilverAndro committed:
Wrong function type, whoops

Hopefully last commit? lmao

OroArmor commented on library/management/chat/src/main/java/org/quiltmc/qsl/chat/api/QuiltChatEvents.java:
```diff
@@ -35,59 +36,34 @@ private QuiltChatEvents() {}
 	 * is recommended.
 	 */
-	public static final ChatEvent<Modify, AbstractChatMessage<?>> MODIFY = new ChatEvent<>(true, (modify, quiltMessageTypes) -> modifyToHook(quiltMessageTypes, modify));
+	public static final ChatEventImpl<Modify, AbstractChatMessage<?>> MODIFY = new ChatEventImpl<>(true, InternalChatEventHandlerConverters::modifyToHook);
```
These should all be `ChatEvent`, not `ChatEventImpl`, the interface should expose everything needed.

OroArmor replied:
```suggestion
	public static final ChatEvent<Modify, AbstractChatMessage<?>> MODIFY = new ChatEventImpl<>(true, InternalChatEventHandlerConverters::modifyToHook);
```

OroArmor commented on library/management/chat/src/main/java/org/quiltmc/qsl/chat/api/QuiltChatEvents.java:
```diff
@@ -35,59 +36,34 @@ private QuiltChatEvents() {}
 	 * An event that allows you to cancel a message by returning true, or false to allow it to continue through.
 	 */
-	public static final ChatBooleanEvent CANCEL = new ChatBooleanEvent();
+	public static final ChatEventBooleanImpl<Cancel> CANCEL = new ChatEventBooleanImpl<>(InternalChatEventHandlerConverters::cancelToHook);
```
```suggestion
	public static final ChatEvent<Cancel, Boolean> CANCEL = new ChatEventBooleanImpl<>(InternalChatEventHandlerConverters::cancelToHook);
```

OroArmor commented on library/management/chat/src/main/java/org/quiltmc/qsl/chat/api/QuiltChatEvents.java:
```diff
@@ -35,59 +36,34 @@ private QuiltChatEvents() {}
 	 * Before (usually) vanilla does any standard processing with this message. Mods may execute other behavior before or after this event.
 	 */
-	public static final ChatEvent<Listen, Void> BEFORE_PROCESS = new ChatEvent<>(false, (listen, quiltMessageTypes) -> listenToHook(quiltMessageTypes, listen));
+	public static final ChatEventImpl<Listen, Void> BEFORE_PROCESS = new ChatEventImpl<>(false, InternalChatEventHandlerConverters::listenToHook);
```
```suggestion
	public static final ChatEvent<Listen, Void> BEFORE_PROCESS = new ChatEventImpl<>(false, InternalChatEventHandlerConverters::listenToHook);
```

OroArmor commented on library/management/chat/src/main/java/org/quiltmc/qsl/chat/api/QuiltChatEvents.java:
```diff
@@ -35,59 +36,34 @@ private QuiltChatEvents() {}
 	 * After (usually) vanilla does any standard processing with this message. Mods may execute other behavior before or after this event.
 	 */
-	public static final ChatEvent<Listen, Void> AFTER_PROCESS = new ChatEvent<>(false, (listen, quiltMessageTypes) -> listenToHook(quiltMessageTypes, listen));
+	public static final ChatEventImpl<Listen, Void> AFTER_PROCESS = new ChatEventImpl<>(false, InternalChatEventHandlerConverters::listenToHook);
```
```suggestion
	public static final ChatEvent<Listen, Void> AFTER_PROCESS = new ChatEventImpl<>(false, InternalChatEventHandlerConverters::listenToHook);
```

OroArmor commented on library/management/chat/src/main/java/org/quiltmc/qsl/chat/api/QuiltChatEvents.java:
```diff
@@ -35,59 +36,34 @@ private QuiltChatEvents() {}
 		AbstractChatMessage<?> handleMessage(AbstractChatMessage<?> abstractMessage);
 	}
 
+	public interface Cancel {
```
```suggestion
	@FunctionalInterface
	public interface Cancel {
```

OroArmor commented on library/management/chat/src/main/java/org/quiltmc/qsl/chat/api/QuiltMessageType.java:
```diff
@@ -21,16 +21,42 @@
 	CLIENT(QuiltMetaMessageType.SIDE),
 
 	// Directionality
+	/**
+	 * Identifies a message that is leaving the side it is on.
+	 */
 	INBOUND(QuiltMetaMessageType.DIRECTION),
+	/**
+	 * Identifies a message that is arriving on the side it is on.
+	 */
```
These comments seem wrong? inbound is leaving?

OroArmor commented on library/management/chat/src/main/java/org/quiltmc/qsl/chat/impl/ChatEventBooleanImpl.java:
```diff
@@ -76,44 +82,27 @@ private boolean matchesMetaTypeRule(QuiltMessageType.QuiltMetaMessageType metaTy
+		backingEvent.register(converter.apply(handler, types));
 	}
 
 	public void addPhaseOrdering(@NotNull Identifier firstPhase, @NotNull Identifier secondPhase) {
```
```suggestion
	@Override
	public void addPhaseOrdering(@NotNull Identifier firstPhase, @NotNull Identifier secondPhase) {
```

OroArmor reviewed (changes requested):
Accidentally approved. Meant to request changes

SilverAndro commented:
whoops, lots of minor mistakes

Guess thats what I get for doing this between classes

SilverAndro committed:
Fix minor mistakes, minor doc changes

SilverAndro committed:
More docs + some tweaks

SilverAndro committed:
These docs are cursed but I think this is how you do it??

Also an NPE fix + clearer naming in a few cases

SilverAndro committed:
Add some more basic docs

SilverAndro committed:
Handler -> Callback

SilverAndro committed:
wow!

SilverAndro committed:
Add CANCELED event

SilverAndro committed:
Final javadoc tweak

EnnuiL commented:
🤦‍♀️ I forgot to add the FCP label here; The FCP began on January 17th, 20:32 (UTC -3)

EnnuiL commented on library/management/chat/src/main/java/org/quiltmc/qsl/chat/api/QuiltChatEvents.java:
```diff
@@ -0,0 +1,88 @@
+	 *
+	 * @see #CANCEL
+	 */
+	public static final ChatEvent<Listen, Void> CANCELED = new ChatEventImpl<>(false, InternalChatEventCallbackConverters::listenToHook);
```
It's actually called `CANCELLED`!

EnnuiL reviewed (changes requested):
oh no

SilverAndro committed:
Took the L there ngl

LambdAurora commented on library/management/chat/src/main/java/org/quiltmc/qsl/chat/api/ChatEvent.java:
```diff
@@ -0,0 +1,91 @@
+ */
+public interface ChatEvent<C, R> {
+	/**
+	 * Invokes the event with the provided message and returns the result or null. A null result is usually the result of no callbacks
```
```suggestion
	 * Invokes the event with the provided message and returns the result or {@code null}. A {@code null} result is usually the result of no callbacks
```

LambdAurora commented on library/management/chat/src/main/java/org/quiltmc/qsl/chat/api/ChatEvent.java:
```diff
@@ -0,0 +1,91 @@
+	 * being attached to this event, but may occur for any other reason depending on the implementation.
+	 *
+	 * @param message the message for the event to process
+	 * @return the result of invoking this event, or null
```
```suggestion
	 * @return the result of invoking this event, or {@code null}
```

LambdAurora commented on library/management/chat/src/main/java/org/quiltmc/qsl/chat/api/ChatEvent.java:
```diff
@@ -0,0 +1,91 @@
+	 * Invokes the event with the provided message, but replaces the result with {@code ifNull} if the result would have been null.
+	 *
+	 * @param message the message for the event to process
+	 * @param ifNull a value that should be returned instead of null
```
```suggestion
	 * @param ifNull a value that should be returned instead of {@code null}
```

LambdAurora commented on library/management/chat/src/main/java/org/quiltmc/qsl/chat/api/ChatEvent.java:
```diff
@@ -0,0 +1,91 @@
+	@Nullable R invoke(@NotNull AbstractChatMessage<?> message);
+
+	/**
+	 * Invokes the event with the provided message, but replaces the result with {@code ifNull} if the result would have been null.
```
```suggestion
	 * Invokes the event with the provided message, but replaces the result with {@code ifNull} if the result would have been {@code null}.
```

LambdAurora commented on library/management/chat/src/main/java/org/quiltmc/qsl/chat/api/ChatEvent.java:
```diff
@@ -0,0 +1,91 @@
+	 * @param firstPhase  the identifier of the phase that should run before the other. It will be created if it didn't exist yet
+	 * @param secondPhase the identifier of the phase that should run after the other. It will be created if it didn't exist yet
+	 */
+	void addPhaseOrdering(@NotNull Identifier firstPhase, @NotNull Identifier secondPhase);
```
I'm wondering if it's possible to have `@Contract("$1, $1 -> throw")` or something in those lines, if it's possible that would be a nice addition.

Leo40Git replied:
Theoretically `@Contract("param1, param1 -> fail")`, but `param<N>` is specifically documented as "The method _returns_ its first (second, third, and so on) argument", so it might not work.
(see https://www.jetbrains.com/help/idea/contract-annotations.html#syntax)

SilverAndro replied:
Can confirm that doesnt work, and I cant find a replacement

LambdAurora replied:
bummer, it could've been super nice

damn you JetBrains!

LambdAurora commented on library/management/chat/src/main/java/org/quiltmc/qsl/chat/api/QuiltChatEvents.java:
```diff
@@ -0,0 +1,88 @@
+	public static final ChatEvent<Modify, AbstractChatMessage<?>> MODIFY = new ChatEventImpl<>(true, InternalChatEventCallbackConverters::modifyToHook);
+
+	/**
+	 * An event that allows you to cancel a message by returning true, or false to allow it to continue through.
```
```suggestion
	 * An event that allows you to cancel a message by returning {@code true}, or {@code false} to allow it to continue through.
```

LambdAurora reviewed (changes requested):
Neat-looking API!

I'd appreciate someone to go over the code again and add `@Contract(pure = true)` where relevant to help IDEs identify pure functions :)

Also whether it's before merge (through a commit in the PR) or during merging, I'd appreciate the addition of `this` keyword in front of class-local attributes that are used to make them more easy to identify.

SilverAndro committed:
Javadoc fixes, method contracts

SilverAndro committed:
`this.sendKeypress(Key.MOD_CTRL | Key.V)`

aka add `this.` to everything

SilverAndro committed:
forgor the impl

SilverAndro committed:
I should really just, double check my stuff before pushing

LambdAurora commented on library/management/chat/src/main/java/org/quiltmc/qsl/chat/api/QuiltChatEvents.java:
```diff
@@ -0,0 +1,88 @@
+	 */
+	@FunctionalInterface
+	public interface Cancel {
+		boolean shouldCancel(AbstractChatMessage<?> abstractMessage);
```
I'm wondering if this method is ok, it having AbstractChatMessage as an argument should reduce conflict risks but the name might be weird in the middle of a listener class.

If my message was confusing: imagine a class that implements this interface and lot of others and add an instance of it to the event, this method name might get confusing in such class.

SilverAndro replied:
Yeah makes sense, ill fix those up

SilverAndro committed:
Bring all mixins in line with naming scheme

LambdAurora commented on library/management/chat/src/main/java/org/quiltmc/qsl/chat/impl/ChatEventBooleanImpl.java:
```diff
@@ -0,0 +1,113 @@
+			// If the hook isn't looking for it
+			if (!hookTypes.contains(messageType)) {
+				// If it doesn't match the complex rule
+				if (!matchesMetaTypeRule(messageType.metaType, hookTypes)) {
```
```suggestion
				if (!this.matchesMetaTypeRule(messageType.metaType, hookTypes)) {
```

LambdAurora commented on library/management/chat/src/main/java/org/quiltmc/qsl/chat/impl/ChatEventImpl.java:
```diff
@@ -0,0 +1,130 @@
+			// If the hook isn't looking for it
+			if (!hookTypes.contains(messageType)) {
+				// If it doesn't match the complex rule
+				if (!matchesMetaTypeRule(messageType.metaType, hookTypes)) {
```
```suggestion
				if (!this.matchesMetaTypeRule(messageType.metaType, hookTypes)) {
```

SilverAndro committed:
`this.`, naming clarity, split event callbacks

Leo40Git reviewed (approved):
WHOOPS, forgot to review again
LGTM!

EnnuiL reviewed (approved):
oh dear, I forgot to re-approve; There we go!

EnnuiL commented on library/management/chat/src/main/java/org/quiltmc/qsl/chat/api/QuiltChatEvents.java:
```diff
@@ -0,0 +1,104 @@
+	public static final ChatEvent<Cancel, Boolean> CANCEL = new ChatEventBooleanImpl<>(InternalChatEventCallbackConverters::cancelToHook);
+
+	/**
+	 * An event that allows you to listen for messages that have been canceled.
```
oh, final nitpick:
```suggestion
	 * An event that allows you to listen for messages that have been cancelled.
```

SilverAndro committed:
More Ls for me

Co-authored-by: Ennui Langeweile <85590273+EnnuiL@users.noreply.github.com>

EnnuiL committed:
Merge branch '1.19.3' into 1.19.3
